### PR TITLE
[SPARK-43204][SQL] Align MERGE assignments with table attributes

### PR DESCRIPTION
**Pull request by aokolnychyi**
### What changes were proposed in this pull request?

This PR extends `ResolveRowLevelCommandAssignments` to also cover MERGE assignments.

### Why are the changes needed?

Similar to SPARK-42151, these changes are needed so that we can rewrite MERGE statements into executable plans for tables that support row-level operations. In particular, our row-level mutation framework assumes Spark is responsible for building an updated version of each affected row and that row is passed back to the data source.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This PR comes with tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1818,7 +1818,10 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             "sqlExpr" -> a.sql,
             "cols" -> cols))
       }
-      resolved
+      resolved match {
```
Comment (aokolnychyi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1821]: Without this change, nested references in MERGE are resolved as `t.struct_col.id AS id`, which breaks our assumption about what kind of keys we expect in assignments. It seems safe for both keys and values but let me know if there are any cases when it is not safe.

Comment (aokolnychyi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1821]: This change is needed as references to nested columns in MERGE are resolved like `a.nested_col.i AS i` (with added aliases), which breaks our assumption about what kind of keys we expect. Let me know if anyone can spot any edge cases when it is not safe.

**Comment by aokolnychyi** [comment]
@cloud-fan @sunchao @viirya @huaxingao @dongjoon-hyun @gengliangwang, this is a follow-up to PR #40308.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @cloud-fan!

**Pull request closed without merging**